cpp/test/sg/genetic/evolution_test.cu

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuml/genetic/common.h>
+#include <cuml/genetic/genetic.h>
+#include <cuml/genetic/node.h>
+#include <cuml/genetic/program.h>
+#include <gtest/gtest.h>
+#include <test_utils.h>
+#include <algorithm>
+#include <cmath>
+#include <cuml/common/logger.hpp>
+#include <iostream>
+#include <raft/handle.hpp>
+#include <rmm/device_uvector.hpp>
+#include <vector>
+
+namespace cuml {
+namespace genetic {
+
+/**
+ * @brief Tests the training and inference of the symbolic regressor, classifier and transformer
+ *        on y = 0.5X[0] + 0.4 X[1]
+ *
+ */
+class GeneticEvolutionTest : public ::testing::Test {
+ protected:
+  void SetUp() override
+  {
+    ML::Logger::get().setLevel(CUML_LEVEL_DEBUG);
```

Comment:
We might want to raise the level here for regular building just to keep the logs clean.

Reply:
> Code looks good. I found mostly minor stuff. Have you done any perf analysis on this compared to the fully CPU version? What's the speedup and scale look like with this implementation?
> 
> Also, I assume at some point this would be exposed through Python, correct?

have pinged you the details regarding the benchmarking Corey. Regarding python, IIUC we are planning to have python bindings, don't think it's on the cards in the near-term future